### PR TITLE
[Feature] Add Lifecycle management for the variables in Context

### PR DESCRIPTION
**Pull request by DavdGao**
## Feature
- As the title says, now we can manage the variables by creating a variable with a wrapper `CtxVar` and a parameter to specify the lifecycle as follows. The variable will be deleted at the end of the corresponding unit. 
```
ctx.loss_batch = CtxVar(ctx.criterion(loss, y_prob), LIFECYCLE.BATCH)
```
- The class `CtxVar` makes no difference to the variables, and you don't need to worry about it. Also, you can manage the lifecycle all by yourself.
- Since all the variables will be deleted at the end of the running routine, we don't need to name the variables with the current mode anymore (e.g., ctx[f"{ctx.cur_mode}_loss_total"]).

## Small modification
- Rename `cur_data_split` into `cur_split` for short
- The function `_run_routine` now returns `num_samples`, which is the number of train/test/validate samples. 

## Note
- After discussion, we postpone the feature of namespace, as for now we didn't see the demand of nested running routines.

### CODE REVIEW

**Review by xieyxclack — COMMENTED** [review]
LGTM

File: federatedscope/core/trainers/context.py [hunk 1]
```diff
@@ -121,21 +117,15 @@ def setup_vars(self):
...
             # local_update_steps
-            num_train_batch, num_train_batch_last_epoch, num_train_epoch, \
-                num_total_train_batch = self.pre_calculate_batch_epoch_num(
-                    self.cfg.train.local_update_steps)
-
-            self.num_train_epoch = num_train_epoch
-            self.num_train_batch = num_train_batch
-            self.num_train_batch_last_epoch = num_train_batch_last_epoch
-            self.num_total_train_batch = num_total_train_batch
+            self.num_train_batch, self.num_train_batch_last_epoch, self.num_train_epoch, self.num_total_train_batch = calculate_batch_epoch_num(self.cfg.train.local_update_steps,
```
[the remark this answers is not in the document]
Comment (DavdGao) [reply federatedscope/core/trainers/context.py]: modified accordingly

File: federatedscope/core/trainers/context.py [hunk 2]
```diff
@@ -84,24 +25,79 @@ def __getattr__(self, item):
...
+        - The variables within an instance of class `Context` can be set/get as an attribute.
+        ```
+        ctx.${NAME_VARIABLE} = ${VALUE_VARIABLE}
+        ```
+        where `${NAME_VARIABLE}` and `${VALUE_VARIABLE}` is the name and value of the variable.
+
+        - To achieve lifecycle management, you can wrap the variable with `CtxVar` and a lifecycle parameter
+        as follows
+        ```
+        ctx.${NAME_VARIABLE} = CtxVar(${VALUE_VARIABLE}, LFECYCLE.BATCH)
```
Comment (yxdyc) [inline federatedscope/core/trainers/context.py:67]: Perhaps we need to add sth. like your mentioned sentence to make users aware that they can also manage the lifecycle by themself if they want, `The class CtxVar makes no difference to the variables, and you don't need to worry about it. Also, you can manage the lifecycle all by yourself.`

**Review by yxdyc — COMMENTED** [review]
LGTM.

File: federatedscope/core/monitors/metric_calculator.py [hunk 3]
```diff
@@ -62,13 +62,13 @@ def _check_and_parse(self, ctx):
             y_prob: The output of the model
 
         """
-        if not '{}_y_true'.format(ctx.cur_data_split) in ctx:
+        if ctx.get('ys_true', None) is None:
             raise KeyError('Missing key y_true!')
```
[the remark this answers is not in the document]
Comment (DavdGao) [reply federatedscope/core/monitors/metric_calculator.py:66]: > missing key ys_true

modified accordingly

File: federatedscope/core/monitors/metric_calculator.py [hunk 4]
```diff
@@ -62,13 +62,13 @@ def _check_and_parse(self, ctx):
...
-        if not '{}_y_prob'.format(ctx.cur_data_split) in ctx:
+        if ctx.get('ys_prob', None) is None:
             raise KeyError('Missing key y_prob!')
```
[the remark this answers is not in the document]
Comment (DavdGao) [reply federatedscope/core/monitors/metric_calculator.py:68]: > similar to above, ys_prob

modified accordingly

File: federatedscope/core/trainers/context.py (hunk 2 above)
Comment (DavdGao) [reply federatedscope/core/trainers/context.py]: > Perhaps we need to add sth. like your mentioned sentence to make users aware that they can also manage the lifecycle by themself if they want, `The class CtxVar makes no difference to the variables, and you don't need to worry about it. Also, you can manage the lifecycle all by yourself.`

Modified accordingly.

**Review by xieyxclack — APPROVED** [review]

**Pull request merged**